docs/proposals/20191016-clusterctl-redesign.md

```diff
@@ -0,0 +1,265 @@
+
+### User Stories
+
+#### Initial Deployment
+
+- As a Kubernetes operator, I’d like to have a simple way for clusterctl to install Cluster API providers into a management cluster, using a very limited set of resources and prerequisites, ideally from a laptop with something I can install with just a couple of terminal commands.
+- As a Kubernetes operator, I’d like to have the ability to provision clusters on different providers, using the same management cluster.
+- As a Kubernetes operator I would like to have a consistent user experience to configure and deploy clusters across cloud providers.
+
+#### Day Two Operations “Lifecycle Management”
+
+- As a Kubernetes operator I would like to be able to install new Cluster API providers into a management cluster.
+- As a Kubernetes operator I would like to be able to upgrade Cluster API components (including CRDs, controllers, etc).  
+- As a Kubernetes operator I would like to have a simple user experience to cleanly remove the Cluster API objects from a management cluster.
+
+#### Target Cluster Pivot Management
+
+- As a Kubernetes operator I would like to pivot Cluster API components from a management cluster to a target cluster.
+
+#### Provider Enablement
+
+- As a Cluster API provider developer, I would like to use my current implementation, or a new version, with clusterctl without recompiling.  
+
+### Implementation Details/Notes/Constraints
+
+> Portions of clusterctl may require changes to the Cluster API data model and/or the definition of new conventions for the providers in order to enable sets of user stories listed above, therefore a roadmap and feature deliverables will need to be coordinated over time.
+
+As of this writing, we envision clusterctl’s component model to look similar to the diagram below:
+
+![components](images/clusterctl-redesign/components.png)
+
+Clusterctl consists of a single binary artifact implementing a CLI; the same features exposed by the CLI will be made available to other management tools via a client library.
+
+During provider installation, clusterctl internal pkg/lib is expected to access provider repositories and read the yaml file specifying all the provider components to be installed in the management cluster.
+
+If the provider repository contains deployment templates, or flavors, clusterctl could be used to generate yaml for creating new workload clusters.  
+
+Behind the scenes clusterctl will apply labels to any and all providers’ components and use a set of custom objects for keeping track of some metadata about installed providers, like e.g. the current version or the namespace where the components are installed. 
+
+#### Init sequence
+
+The clusterctl CLI is optimized for the “day one” experience of using Cluster API, and it will be possible to create a first cluster with two commands after the pre-requisites are met:
+
+```shell
+$ clusterctl init --infrastructure aws
```

Comment:
Will `init` allow user to provide specific version of a provider?

Reply:
TBD.
For the default behavior, the version is encoded in the provider configuration (either a specific version or "latest"); users, will be allowed to override such configuration via a config file.

Once this is in place, we can discuss if/how to allow the user to provide specific versions via command line as well, but this requires some additional consideration about the UX because in init many providers can be added at the same time.

Reply:
after some discussions, we are going to support this from CLI as well.

Probably the syntax will be docker like
```
clusterctl init --infrastructure aws:v0.5.3
```
But exact details will be defined during implementation